File: server/utils/twitter_sentiment.js
import dotenv from 'dotenv';
import { TwitterApi } from 'twitter-api-v2';
import Sentiment from 'sentiment';
import fetch from 'node-fetch'; // Required for any Groq-style AI fallback

dotenv.config();

const BEARER_TOKEN = process.env.TWITTER_BEARER_TOKEN;
const GROQ_API_KEY = process.env.GROQ_API_KEY;

if (!BEARER_TOKEN) {
  console.error("❌ ERROR: Missing Twitter Bearer Token in .env");
  process.exit(1);
}

if (!GROQ_API_KEY) {
  console.error("❌ ERROR: Missing Groq AI Key in .env");
  process.exit(1);
}

const twitterClient = new TwitterApi(BEARER_TOKEN);
const sentiment = new Sentiment();

function analyzeSentiment(text) {
  const result = sentiment.analyze(text);
  const score = result.score;

  if (score > 0) return 'positive';
  if (score < 0) return 'negative';
  return 'neutral';
}

async function getLiveTweets(keyword) {
  try {
    const query = `${keyword} -is:retweet lang:en`;

    const tweetsResponse = await twitterClient.v2.search(query, {
      max_results: 10,
    });

    if (!tweetsResponse || !tweetsResponse.data?.data?.length) {
      console.warn(`⚠️ No tweets found for: ${keyword}`);
      return JSON.stringify([]);
    }

    const sentiments = tweetsResponse.data.data.map(tweet => ({
      tweet: tweet.text,
      sentiment: analyzeSentiment(tweet.text)
    }));

    return JSON.stringify(sentiments, null, 2);

  } catch (error) {
    if (error.code === 429) {
      console.warn("⚠️ Rate limit exceeded. Using Groq AI to generate tweets...");
      // Optional: Fallback logic to use Groq/OpenAI API if rate limited
    } else {
      console.error("❌ ERROR:", error.message);
    }

    return JSON.stringify([]);
  }
}

// Run (only if called directly)
const keyword = "NFT";
if (process.argv[1] === new URL(import.meta.url).pathname) {
  getLiveTweets(keyword).then(console.log);
}

// ✅ Export for external usage
export { getLiveTweets };

File: server/utils/twitter_sentiment.py
import tweepy
import os
import json
import time
from dotenv import load_dotenv
from textblob import TextBlob
import sys
import openai  # Groq AI uses OpenAI-compatible API

sys.stdout.reconfigure(encoding='utf-8')  # Force UTF-8 support

# Load environment variables
load_dotenv()

BEARER_TOKEN = os.getenv("TWITTER_BEARER_TOKEN")
GROQ_API_KEY = os.getenv("GROQ_API_KEY")

if not BEARER_TOKEN:
    print("ERROR: Missing Twitter Bearer Token in .env")
    exit(1)

if not GROQ_API_KEY:
    print("ERROR: Missing Groq AI Key in .env")
    exit(1)

client = tweepy.Client(bearer_token=BEARER_TOKEN)

def analyze_sentiment(text):
    """Returns sentiment label based on TextBlob analysis."""
    analysis = TextBlob(text)
    polarity = analysis.sentiment.polarity
    return "positive" if polarity > 0 else "negative" if polarity < 0 else "neutral"

def get_live_tweets(keyword):
    """Fetches tweets from Twitter API and analyzes sentiment."""
    try:
        query = f"{keyword} -is:retweet lang:en"
        tweets = client.search_recent_tweets(query=query, max_results=10)

        if not tweets.data:
            print(f"⚠️ No tweets found for: {keyword}")
            return json.dumps([])

        sentiments = []
        for tweet in tweets.data:
            sentiments.append({
                "tweet": tweet.text,
                "sentiment": analyze_sentiment(tweet.text)
            })

        return json.dumps(sentiments)

    except tweepy.TooManyRequests:
        print("⚠️ Rate limit exceeded. Using Groq AI to generate tweets...")

    except Exception as e:
        print(f"ERROR: {str(e)}")
        return json.dumps([])

if __name__ == "__main__":
    keyword = "NFT"
    print(get_live_tweets(keyword))
